File: web/src/rows.js
// Shared client-side presentation layer for entity rows (search results, thesaurus reflex list,
// reconstructions index). Goal: ONE row builder and ONE url builder per entity type, so the CLIENT
// views can't drift — a reflex row used to be hand-built in three places and quietly diverged (see
// the project-attestation-links convention). Pure + Node-importable (no DOM at import; B falls back
// to '' off-browser).
//
// SYNC — the static pages (language / etymon / group / thesaurus) are rendered SERVER-side in Python
// and do NOT use this file, so several builders below have a Python twin that must render the SAME
// markup. When you change one, change its twin: both sides are tagged `SYNC(<key>)` — grep the key to
// find every site. Twins: SYNC(reflex-row), SYNC(etymon-row), SYNC(syllabify), SYNC(syllable-links),
// SYNC(protoform-fmt), SYNC(entity-urls).

export const B = (typeof window !== 'undefined' && window.STEDT_BASE) || '';
export const esc = s => String(s == null ? '' : s).replace(/[&<>"]/g, c => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));
// SYNC(protoform-fmt) ↔ stedt/render/text.py alt() — normalise a proto-form (strip the leading *,
// which every emission site re-adds as literal text; star each ⪤-alternant). Keep identical.
export const altstar = s => String(s).replace(/^\s*\*\s*/, '').replace(/(⪤|\bOR\b|~|=)\s*\*?/g, '$1 *');
export const fmt = n => Number(n).toLocaleString();
// SYNC(sortkey) ↔ text.py sortkey + search.js sortkey — the case/accent-insensitive collation key
export const norm = s => String(s == null ? '' : s).toLowerCase().normalize('NFD').replace(/[̀-ͯ]/g, '');

// --- canonical URLs: the ONE place each entity's address is built ---
// SYNC(entity-urls) ↔ stedt/render/shell.py {etymon,source,language,reflex}_href + the /thesaurus/{sk}
// category link — the server builds the same addresses; keep the URL shapes + conventions identical.
export const languageHref = lgid => `${B}/language/${lgid}`;            // top of a language page
export const reflexHref = (lgid, rn) => `${B}/language/${lgid}#rn${rn}`; // a specific attestation row
export const etymonHref = tag => `${B}/etymon/${tag}`;
export const sourceHref = abbr => `${B}/source/${esc(abbr)}`;
export const categoryHref = sk => `${B}/thesaurus/${esc(sk)}`;       // a reflex's semantic-category node
// render_note (server) ships root-relative xref links (/etymon/…) in the search DB's note HTML;
// prepend the page base so they resolve under /stedt (or wherever the site is mounted). The note is
// then injected as HTML (it's already escaped/sanitised by render_note), not re-escaped.
export const rebase = html => String(html == null ? '' : html).replace(/href="\//g, `href="${B}/`);

// --- per-syllable etymon links (faithful port of the original SylStation.syllabify): when a
// reflex's syllables are individually tagged, each links to its etymon. Char classes [(] [)] [|]
// stand in for the escaped \( \) \| to keep this readable. ---
// SYNC(syllabify) ↔ stedt/render/syllabify.py — same tokenizer in two runtimes; verified byte-equal
// over 40k tagged forms. Any change must stay identical (re-run that diff).
const _TONE = "⁰¹²³⁴⁵⁶⁷⁸0-9ˊˋ˥-˩";
const _DELIM = "-=≡≣+.,;/~◦⪤()↮ ";
const _HIDE = new RegExp('[(]([^' + _DELIM + _TONE + ']+)[)]', 'g');
const _START = new RegExp('^([' + _DELIM + ']+)');
const _REPOST = "([^" + _DELIM + _TONE + "]+[" + _TONE + "]+(?:[|]$)?)([" + _DELIM + "]*)";
const _REPRE = "([" + _TONE + "]{1,2}[^" + _DELIM + _TONE + "]+)([" + _DELIM + "]*)";
const _REDEL = "([^" + _DELIM + "]+)([" + _DELIM + "]*)";
function _syl1(s, reSrc) {
  s = s.replace(_HIDE, '（$1）'); let prefix = '';
  if (_START.test(s)) { const pm = _START.exec(s); prefix = pm[1]; s = s.substring(prefix.length); }
  const syls = [], dl = []; const re = new RegExp("^" + reSrc); let m;
  while ((m = re.exec(s)) && m[0].length) {
    s = s.substring(m[0].length);
    if (m[1].indexOf('|') !== -1 && syls.length) {
      syls[syls.length - 1] += dl.pop();
      syls[syls.length - 1] += m[1].replace(/（/g, '(').replace(/）/g, ')').replace('|', '');
    } else { syls.push(m[1].replace(/（/g, '(').replace(/）/g, ')')); }
    dl.push(m[2]);
  }
  if (!syls[0]) syls[0] = '';
  if (s) syls[syls.length - 1] += s;
  return { syls, dl, prefix, ok: !s.length };
}
function syllabify(s) {
  let r = _syl1(s, _REPOST);
  if (!r.ok) { r = _syl1(s, _REPRE); if (!r.ok) r = _syl1(s, _REDEL); }
  return r;
}
// SYNC(syllable-links) ↔ stedt/render/rows.py syl_pop(): the linked syllable's etymon-preview
// popover — the original's elink popup, all three parts: linked '#tag PLG *protoform ‘gloss’'
// header, the etymon's mesoroots (each PLG linked to its #ms-{grpno} row on the etymon page),
// and the allofam family (members linked, this etymon bold). The card sits BESIDE the trigger
// link inside .syl-w (nested <a> is invalid), so its links are real. Keep the markup identical.
const sylPop = e => {
  const g = e.pg ? ` ‘${esc(e.pg)}’` : '';
  let out = `<a class="sp-h" href="${etymonHref(e.tag)}">#${e.tag}${e.plg ? ' ' + esc(e.plg) : ''} *${altstar(esc(e.pf))}${g}</a>`;
  const meso = Array.isArray(e.meso) ? e.meso : [], fam = Array.isArray(e.fam) ? e.fam : [];
  if (meso.length) {
    if (fam.length) out += '<span class="sp-sec">Mesoroots</span>';
    out += meso.map(m =>
      `<a class="sp-m" href="${etymonHref(e.tag)}#ms-${esc(String(m.no || ''))}"><span class="sp-plg">${esc(m.plg || '')}</span> *${altstar(esc(m.f))}${m.g ? ` ‘${esc(m.g)}’` : ''}</a>`
    ).join('');
  }
  if (fam.length) {
    out += '<span class="sp-sec">Allofams</span>';
    out += fam.map(a => {
      const lab = `${esc(a.s)} #${a.tag}${a.plg ? ' ' + esc(a.plg) : ''} *${altstar(esc(a.pf))}${a.pg ? ` ‘${esc(a.pg)}’` : ''}`;
      return a.tag === e.tag
        ? `<span class="sp-m"><b>${lab}</b></span>`
        : `<a class="sp-m" href="${etymonHref(a.tag)}">${lab}</a>`;
    }).join('');
  }
  return `<span class="sylpop">${out}</span>`;
};

// SYNC(morph-codes) ↔ stedt/render/text.py morph_code/morph_label — STEDT's per-morpheme analysis
// codes (the original's lexicon.analysis): each non-cognate morpheme slot is tagged 'b' (borrowing,
// optionally with a source, e.g. 'bIndic'), 'p' (prefix), 's' (suffix), or 'm' (an identifiable but
// unreconstructed morpheme). A code is a run of letters/'?' only. Keep the predicate + labels
// identical to the server twin so both runtimes mark the same morphemes the same way.
const MORPH_RE = /^[A-Za-z?]+$/;
export const morphCode = tok => (tok && MORPH_RE.test(tok)) ? tok : null;
const morphLabel = code => {
  if (code === 'p') return 'prefix';
  if (code === 's') return 'suffix';
  if (code === 'm') return 'morpheme';
  if (code[0] === 'b') {                  // borrowing: b / b? / bSOURCE / b?SOURCE / bSOURCE?
    let rest = code.slice(1);
    const uncertain = rest.startsWith('?') || rest.endsWith('?');
    rest = rest.replace(/^\?+|\?+$/g, '');
    const lab = rest ? `${rest} loanword` : 'loanword';
    return uncertain ? `probable ${lab}` : lab;
  }
  return code;
};
// SYNC(morph-codes) ↔ stedt/render/rows.py morph_mark — a coded morpheme: text marked (.morph;
// borrowings add .morph-b) with a popover BESIDE it inside .morph-w (mirrors .syl-w). base escaped.
const morphMark = (code, base) =>
  `<span class="morph-w"><span class="${code[0] === 'b' ? 'morph morph-b' : 'morph'}">${base}</span><span class="mpop">${esc(morphLabel(code))}</span></span>`;
// SYNC(reflex-row) ↔ stedt/render/rows.py morph_chip — fallback trailing summary of the codes,
// used when the form can't be syllabified so the marks can't sit on the morphemes themselves.
const morphChip = codes => {
  if (!codes) return '';
  const labs = Object.keys(codes).map(Number).sort((a, b) => a - b).map(i => esc(morphLabel(codes[i]))).join(' · ');
  return `<span class="anl morphs">${labs}</span>`;
};

// SYNC(syllable-links) ↔ stedt/render/rows.py syl_form() + syl_pop(): the syllable-linked form
// + its etymon-preview popover. Keep the markup (a.syl, .sylpop, header + mesoroot lines) identical.
const sylLink = r => {                     // syllable-linked form HTML, or null to fall back
  if (!r.syn && !r.morph) return null;
  const syn = r.syn || {}, morph = r.morph || {};
  // headword arrives as `form` (search payload) or `reflex` (category payload) — same fallback
  // as the plain path below, or category rows lose their links (and ind-0 rows lost their TEXT:
  // syllabifying '' yields one empty syllable, which happily took the link)
  const head = r.form != null ? r.form : r.reflex;
  const sy = syllabify(String(head || '')), syls = sy.syls, dl = sy.dl;
  for (const k in syn) { if (+k >= syls.length) return null; }     // cognate tags must land on real syllables
  for (const k in morph) { if (+k >= syls.length) return null; }   // and so must codes
  const info = {};                         // tag -> {pf, pg}, for the syllable's etymon-preview popover
  (r.etyma || []).forEach(e => { if (e && e.tag != null) info[e.tag] = e; });
  let out = esc(sy.prefix || '');
  for (let i = 0; i < syls.length; i++) {
    const tag = syn[i];
    if (tag != null) {
      const base = esc(syls[i]), e = info[tag];
      const pop = e && e.pf ? sylPop(e) : '';
      const link = `<a class="syl" href="${etymonHref(tag)}">${base}</a>`;
      out += pop ? `<span class="syl-w">${link}${pop}</span>` : link;
    } else if (morph[i] != null) {
      out += morphMark(morph[i], esc(syls[i]));
    } else {
      out += esc(syls[i]);
    }
    out += esc(dl[i] || '').replace(/◦/g, '<span class="br">◦</span>');
  }
  return out;
};

// SYNC(display-form) ↔ stedt/render/rows.py disp_form — plain (unlinked) display of a stored form:
// strip the internal '|' analysis delimiter via the same syllabify+rejoin the linked path uses,
// escape, and mute the ◦ morpheme separator. Unpiped forms pass straight through.
export const dispForm = s => {
  s = String(s == null ? '' : s);
  if (s.indexOf('|') < 0) return esc(s).replace(/◦/g, '<span class="br">◦</span>');
  const sy = syllabify(s);
  let out = esc(sy.prefix || '');
  for (let i = 0; i < sy.syls.length; i++) {
    out += esc(sy.syls[i]);
    out += esc(sy.dl[i] || '').replace(/◦/g, '<span class="br">◦</span>');
  }
  return out;
};

// --- entity rows ---

// A reflex, shared by the search results and the thesaurus reflex list so they can't drift.
// SYNC(reflex-row) ↔ the server-rendered reflex rows in stedt/render/language.py — language()
// seginfo + stedt/render/etymon.py etymon() rfx builder. Keep the fields, order (POS before gloss), classes,
// roles (listitem — every container holding these rows carries role=list), and link
// targets identical across both runtimes.
// The whole row links to the form's attestation line (#rn) via a stretched overlay; the inner links
// sit above it (see .rx-go in site.css): the language name → the TOP of its language page, syllables
// / via chips → their etyma, source → its page, and a noted gloss stays interactive (shows its note).
// Accepts either `form` (search payload) or `reflex` (category payload) for the headword.
export const reflexRow = r => {
  const loc = r.srcid ? `: ${esc(r.srcid)}` : '';   // per-reflex locus (page/entry), like the entity pages
  const src = r.srcabbr ? `<a href="${sourceHref(r.srcabbr)}">${esc(r.citation || r.srcabbr)}${loc}</a>` : '';
  const pos = r.gfn ? `<span class="pos">${esc(r.gfn)}</span>` : '';   // sits before the gloss (.pos has margin-right)
  const gl = r.note
    ? `<span class="g noted" tabindex="0" aria-describedby="np${r.rn}">${esc(r.gloss)}<span class="notepop" role="note" id="np${r.rn}">${rebase(r.note)}</span></span>`
    : `<span class="g">${esc(r.gloss)}</span>`;
  const lf = sylLink(r); let mid;
  // etyma already linked inline in the syllable form don't repeat as trailing chips
  const inline = (lf && r.syn) ? new Set(Object.values(r.syn)) : new Set();
  const vias = (r.etyma || []).filter(x => x && x.tag != null && !inline.has(x.tag))
    .map(x => `<a class="via" href="${etymonHref(x.tag)}">*${altstar(esc(x.pf))}</a>`);
  const links = vias.length ? ` <span class="vias">${vias.join(' ')}</span>` : '';
  if (lf) {                              // syllables carry their own etymon links / morpheme marks
    mid = `<span class="lat">${lf}</span> ${pos}${gl}${links}`;
  } else {                              // plain form; trailing "via" chips + morpheme codes
    const form = r.form != null ? r.form : r.reflex;
    mid = `<span class="lat">${dispForm(form)}</span> ${pos}${gl}${links}${morphChip(r.morph)}`;
  }
  // the reflex's semantic category (search rows only; the thesaurus-category list omits it as redundant)
  if (r.cat) mid += ` <a class="rx-cat" href="${categoryHref(r.semkey)}">${esc(r.cat)}</a>`;
  const go = `<a class="rx-go" href="${reflexHref(r.lgid, r.rn)}" aria-label="${esc(r.language)}: go to this entry"></a>`;
  return `<div class="rx-hit" role="listitem">${go}<a class="lang" href="${languageHref(r.lgid)}">${esc(r.language)}</a><span class="rx-mid">${mid}</span><span class="rx-src">${src}</span></div>`;
};

// A reconstruction (etymon) result row.
// SYNC(etymon-row) ↔ the server-rendered etymon lists: stedt/render/group.py reconinfo() (group
// page) + stedt/render/indexes.py dinfo (thesaurus). Keep protoform / PLG / #tag / reflex-count /
// exemplary-badge identical.
export const etymonRow = e => `<a class="ety-hit" href="${etymonHref(e.tag)}"><span class="pf2 lat"><span class="star">*</span>${altstar(esc(e.protoform))}</span><span class="pg2">${esc(e.protogloss)}</span><span class="tagn">${esc(e.plg)} #${e.tag}${e.nreflex ? ` · ${fmt(e.nreflex)} reflex${e.nreflex == 1 ? '' : 'es'}` : ''}${e.exemplary ? ' · <span class="exm">exemplary</span>' : ''}${e.public === 0 || e.public === '0' || e.provisional ? ' · <span class="prov">provisional</span>' : ''}</span></a>`;

// A language result row.
export const languageRow = l => `<a class="ety-hit" href="${languageHref(l.lgid)}"><span class="rf">${esc(l.language)}</span><span class="gl2">${fmt(l.n)} reflex${l.n == 1 ? '' : 'es'}</span><span class="tagn">language</span></a>`;

File: stedt/render/rows.py
"""Shared row fragments used by more than one entity page — the server-side counterpart of
web/src/rows.js (the client renders the same row shapes)."""

from .notes import note_span
from .shell import etymon_href, source_href
from .syllabify import syllabify
from .text import esc, alt, morph_label, seq_label


def syl_pop(tag, info):
    """SYNC(syllable-links) ↔ web/src/rows.js sylPop. The hover/focus popover for a linked
    syllable — the original's elink popup (rootcanal tt/et_info.tt), all three parts:
    - header "#tag PLG *protoform ‘gloss’", linked to the etymon page (the original's '#232:');
    - the etymon's mesoroots ("PLG *form ‘gloss’", Stammbaum-ordered), each PLG linked to its
      row on the etymon page (#ms-{grpno} — the original linked /etymon/tag#grpno);
    - the computed allofam family ("1a #34 PLG *form ‘gloss’"), members linked, the popover's
      own etymon bold and unlinked, exactly the original's list. The original's tab SWITCHER
      is the one omission (tabs can't operate in a hover card): both sections show, labeled.
    The card sits NEXT TO the trigger link (inside .syl-w), not inside it, so its links are
    real. info: (pf, pg, plg, mesoroots, family)."""
    pfx, pgl, plg, meso, fam = info
    g = f" ‘{esc(pgl)}’" if pgl else ""
    head = (f'<a class="sp-h" href="{etymon_href(tag)}">#{tag}{" " + esc(plg) if plg else ""}'
            f' *{esc(alt(pfx))}{g}</a>')
    out = head
    if meso:
        if fam:
            out += '<span class="sp-sec">Mesoroots</span>'
        out += "".join(
            f'<a class="sp-m" href="{etymon_href(tag)}#ms-{esc(str(no or ""))}">'
            f'<span class="sp-plg">{esc(mp or "")}</span> *{esc(alt(mf))}{f" ‘{esc(mg)}’" if mg else ""}</a>'
            for mp, mf, mg, no in meso
        )
    if fam:
        out += '<span class="sp-sec">Allofams</span>'
        for seq, t2, plg2, pf2, pg2 in fam:
            lab = (f'{esc(seq_label(seq))} #{t2}{" " + esc(plg2) if plg2 else ""} *{esc(alt(pf2))}'
                   f'{f" ‘{esc(pg2)}’" if pg2 else ""}')
            out += (f'<span class="sp-m"><b>{lab}</b></span>' if t2 == tag
                    else f'<a class="sp-m" href="{etymon_href(t2)}">{lab}</a>')
    return f'<span class="sylpop">{out}</span>'


def morph_mark(code, base):
    """SYNC(morph-codes) ↔ web/src/rows.js morphMark — a coded (non-cognate) morpheme: the syllable
    text marked (.morph; borrowings add .morph-b to stand out a touch more) with a hover/focus
    popover naming the analysis (text.morph_label). The popover sits BESIDE the mark inside .morph-w,
    mirroring the cognate links' .syl-w, so the two annotation kinds lay out the same. base is already
    escaped."""
    cls = "morph morph-b" if code[:1] == "b" else "morph"
    return f'<span class="morph-w"><span class="{cls}" tabindex="0">{base}</span><span class="mpop">{esc(morph_label(code))}</span></span>'


def morph_chip(codes):
    """SYNC(reflex-row) ↔ web/src/rows.js morphChip — the fallback trailing summary of a reflex's
    morpheme codes (' · prefix · Indic loanword'), used when the form can't be syllabified so the
    marks can't sit on the morphemes themselves. codes: {position: code}, rendered in order."""
    if not codes:
        return ""
    labs = " · ".join(esc(morph_label(codes[i])) for i in sorted(codes))
    return f'<span class="anl">{labs}</span>'


def syl_form(reflex, syn, pf=None, self_tag=None, codes=None):
    """SYNC(syllable-links) ↔ web/src/rows.js sylLink — keep the markup identical.
    Reflex surface form as HTML with each tagged syllable linked to its own etymon (cognate links,
    each carrying a *protoform 'gloss' popover) and each coded morpheme marked (morph_mark). On an
    etymon page, pass self_tag = that etymon: the syllable that IS this etymon is marked but not
    linked (you're already here). Returns None to fall back to the plain form + trailing chips
    (morph_chip carries the codes there). pf: tag -> (protoform, protogloss); codes: {position: code}."""
    syn = syn or {}
    codes = codes or {}
    if not syn and not codes:
        return None
    syls, dl, prefix = syllabify(reflex or "")
    if any(k >= len(syls) for k in syn) or any(k >= len(syls) for k in codes):   # a tag/code must land on a real syllable
        return None
    pf = pf or {}
    out = esc(prefix)
    for i, syl in enumerate(syls):
        tag = syn.get(i)
        base = esc(syl)
        if tag is not None and tag == self_tag:
            out += f'<span class="syl-self">{base}</span>'   # this etymon's own reflex syllable
        elif tag is not None and tag in pf:
            # pf membership gates the link: pf comes from proto_labels(), which is restricted to
            # non-DELETE etyma — the only ones with a built page. A tag outside pf (withdrawn
            # etymon) renders as plain text, matching the client, whose DELETE-filtered join never
            # sees such tags at all.
            # the popover sits BESIDE the link inside a .syl-w wrapper (hover/focus-within on the
            # wrapper reveals it): nested <a> is invalid HTML, and the card now carries links
            info = pf[tag]
            pop = syl_pop(tag, info) if info else ""
            link = f'<a class="syl" href="{etymon_href(tag)}">{base}</a>'
            out += f'<span class="syl-w">{link}{pop}</span>' if pop else link
        elif i in codes:
            out += morph_mark(codes[i], base)
        else:
            out += base
        d = dl[i] if i < len(dl) else ""
        out += esc(d).replace("◦", '<span class="br">◦</span>')
    return out


def disp_form(s):
    """SYNC(display-form) ↔ web/src/rows.js dispForm — the plain (unlinked) rendering of a stored
    form. The '|' in a stored form is STEDT's internal analysis-override delimiter, never part of
    the citation form: the original stripped it from every displayed form, and our syllable-linked
    path does too (syllabify glues 'g-|lak' → 'g-lak'). Route piped forms through the same
    syllabify+rejoin; unpiped forms pass straight through. Escapes, and mutes the ◦ morpheme
    separator like the linked path."""
    s = s or ""
    if "|" not in s:
        return esc(s).replace("◦", '<span class="br">◦</span>')
    syls, dl, prefix = syllabify(s)
    out = esc(prefix)
    for i, syl in enumerate(syls):
        out += esc(syl)
        out += esc(dl[i] if i < len(dl) else "").replace("◦", '<span class="br">◦</span>')
    return out


def lgab_span(lgabbr):
    """The language-abbreviation chip (leading space + <span class="lgab">), or "" when absent."""
    return f' <span class="lgab">{esc(lgabbr)}</span>' if lgabbr else ""


def as_abbr(lgabbr, name):
    """The "as SHORTNAME" connective — a muted .asab span around the .lgab chip — for the lead of
    an entity row's metadata (.subg) column. Empty when the abbr is absent or merely echoes the lect
    name (no point printing "Khaling as Khaling"). SYNC: source.py langinfo ↔ language.py variants
    both lead .subg with this, so a source's short name for a lect reads the same from either end."""
    ab = (lgabbr or "").strip()
    if not ab or ab.casefold() == (name or "").strip().casefold():
        return ""
    return f'<span class="asab">as <span class="lgab">{esc(ab)}</span></span>'


def etymon_flags(exemplary, public, prov_title=None):
    """SYNC(etymon-row) ↔ web/src/rows.js etymonRow — the ' · exemplary' / ' · provisional'
    chips trailing an etymon's tag line, identical wherever an etymon row renders. prov_title:
    the etymon page's own header explains the provisional mark with a tooltip; rows don't."""
    out = ' · <span class="exm">exemplary</span>' if (exemplary or "") == "x" else ""
    if not public:
        t = f' title="{esc(prov_title)}"' if prov_title else ""
        out += f' · <span class="prov"{t}>provisional</span>'
    return out


def noted_gloss(rn, gloss, notes):
    """SYNC(reflex-row) ↔ web/src/rows.js reflexRow — the gloss cell when the reflex carries
    lexical notes: the note popover rides on the gloss behind a circled-i. The popover lives
    INSIDE the inline gloss <span>, so each note renders as an inline note_span (a block <p>
    here would be invalid markup); aria-describedby ties the notes to the gloss for AT.
    notes: [(notetype, xmlnote), …] as lexical_notes() returns them."""
    pop = "".join(note_span(nt, x) for nt, x in notes)
    return (
        f'<span class="g noted" tabindex="0" aria-describedby="np{rn}">{esc(gloss)}'
        f'<span class="notepop" role="note" id="np{rn}">{pop}</span></span>'
    )


def src_cell(srcabbr, citation, srcid):
    """SYNC(reflex-row) ↔ web/src/rows.js reflexRow — the source cell: the work the form is
    attested in (citation linked to its source page) + ': locus' (page/entry/note) within it;
    a record with no source keeps its citation text as a plain span."""
    loc = f": {esc(srcid)}" if srcid else ""
    if srcabbr:
        return f'<a class="src" href="{source_href(srcabbr)}">{esc(citation or srcabbr)}{loc}</a>'
    return f'<span class="src">{esc(citation or "")}{loc}</span>'
